revert to condense from enforce, passing symmetric solver

gdmcbain opened the pull request:
This takes ex21 back to where it was before #591 and #596, thereby avoiding having to excise the spurious imaginary part introduced by the asymmetric eigensolver as in #611 or #612.

Eventually it might be possible to `enforce` rather than condense the Dirichlet conditions, but I don't think it's quite clear yet how to do this for generalized eigenvalue problems.

kinnala commented:
I did not observe any spurious imaginary part, just a different dtype with zero imaginary part, but I think you are correct here in the sense that this is the proper way to solve a symmetric problem.

I changed the behavior of `solve` because I think it should deal with the most general case by default or otherwise the result can be surprising: wrong results without any warning.

gdmcbain commented:
> I did not observe any spurious imaginary part, just a different dtype with zero imaginary part, 

Ah, right, I didn't check that the spurious imaginary part was actually zero.

> but I think you are correct here in the sense that this is the proper way to solve a symmetric problem.

Yeah, there's still all the other problems of #75 and I haven't totally decided that `enforce` can't work for something like this if used correctly; we'll see whether its original proponent chimes in.  It will introduce zero eigenvalues for the Dirichlet degrees of freedom but that doesn't necessarily matter.  I hadn't at first realized that #596 was only replacing the rows and not the columns.  I think in elementary presentations (reference?) the column is also replaced so that symmetry is preserved but of course that's too inefficient for a `csr_matrix` so symmetry is sort-of abandoned but not really as described by E. & G. &sect;8.4.3, &c., &c.

So I think this PR is the right thing to do in the short term (get the example running correctly without warnings or errors or makeshifts), but it might be bettered.

> I changed the behavior of solve because I think it should deal with the most general case by default or otherwise the result can be surprising: wrong results without any warning.

Yes, this was the right thing to do, for sure.
